src/hooks/useQuery.ts

```diff
@@ -51,23 +61,23 @@ export const useQuery = <T = any, V = object>(
             requestPolicy: args.requestPolicy,
             ...opts,
           }),
-          subscribe(({ data, error }) =>
-            setState({ fetching: false, data, error })
-          )
+          subscribe(({ data, error }) => {
+            setState({ fetching: false, data, error });
+          })
         );
       } else {
         setState(s => ({ ...s, fetching: false }));
       }
 
-      unsubscribe = teardown;
+      unsubscribe.current = teardown;
     },
-    [request.key, args.skip, args.requestPolicy]
+    [request, client, args.skip, args.requestPolicy]
```

Comment:
`client` and `request` were both flagged as needing to be included here because they're both accessed in this `useCallback`. `client` comes out of context so I think it should be included for safety. It _should_ only ever get set once, so it wasn't affecting behavior at all previously because it never changed.

Reply:
Ah, damn, ok this is a huge issue and actually introduces a bug. The `request.key` is the hash of the request, but we're recreating the instance on every mount on purpose, since we can't know if a `DocumentNode` or `variables` instance has changed.

Reply:
Ah ok. Does that not get covered by the `useMemo` above, which gets recalculated whenever `args.query` or `args.variables` changes? I guess the way I see the dependencies working is.

`useMemo` – creates `request`, recalculates whenever `args.query` (`DocumentNode`) or `args.variables` (`variables`) changes.
`useCallback` – has a dependency on `request` generated by `useMemo`, so it _should_ get recalculated whenever `args.query` or `args.variables` changes (which sounds like is the case for which a new hash gets generated into `request.key`).
`useEffect` – gets called whenever `executeQuery` (the `useCallback`) gets recalculated.

Just want to make sure I really grok the problem, as the hooks dependencies seem correct. Happy to change back to `request.key` tho!

Reply:
Not really, because we don't want to rely on reference equality here. Variables are likely to be recreated by the user and forcing them to memo those can be a burden. That should be an optimisation that they can optionally apply.

The same goes for misused graphql- tag stuff, where reference equality is often guaranteed but not always

Reply:
Gotcha, I'll ensure we use `request.key` here and we can disable the hooks lint warning.